src/jdk.incubator.foreign/share/classes/jdk/internal/foreign/SystemLookup.java

```diff
@@ -45,29 +53,129 @@ private SystemLookup() { }
+            int numSymbols = WindowsFallbackSymbols.values().length;
+            MemorySegment funcs = fallbackLibLookup.lookup("funcs").orElseThrow()
+                .asSegment(C_POINTER.byteSize() * numSymbols, ResourceScope.newImplicitScope());
+
+            SymbolLookup fallbackLookup = name -> Optional.ofNullable(WindowsFallbackSymbols.BY_NAME.get(name))
+                .map(symbol -> MemoryAccess.getAddressAtIndex(funcs, symbol.ordinal()));
+
+            final SymbolLookup finalLookup = lookup;
+            lookup = name -> finalLookup.lookup(name).or(() -> fallbackLookup.lookup(name));
         }
-    };
+
+        return lookup;
+    }
+
+    private static SymbolLookup libLookup(Function<NativeLibraries, NativeLibrary> loader) {
+        NativeLibrary lib = loader.apply(NativeLibraries.rawNativeLibraries(SystemLookup.class, false));
+        return name -> {
+            Objects.requireNonNull(name);
+            try {
+                long addr = lib.lookup(name);
+                return addr == 0 ?
+                        Optional.empty() : Optional.of(MemoryAddress.ofLong(addr));
+            } catch (NoSuchMethodException e) {
+                return Optional.empty();
+            }
+        };
+    }
 
     @Override
     public Optional<MemoryAddress> lookup(String name) {
-        Objects.requireNonNull(name);
-        long addr = syslookup.find(name);
-        return addr == 0 ?
-                Optional.empty() : Optional.of(MemoryAddress.ofLong(addr));
+        return syslookup.lookup(name);
     }
 
     public static SystemLookup getInstance() {
         return INSTANCE;
     }
+
+    // fallback symbols missing from ucrtbase.dll
+    // this list has to be kept in sync with the table in the companion native library
+    private enum WindowsFallbackSymbols {
+        // stdio
+        fprintf,
+        fprintf_s,
+        fscanf,
+        fscanf_s,
+        fwprintf,
+        fwprintf_s,
+        fwscanf,
+        fwscanf_s,
+        printf,
+        printf_s,
+        scanf,
+        scanf_s,
+        snprintf,
+        sprintf,
+        sprintf_s,
+        sscanf,
+        sscanf_s,
+        swprintf,
+        swprintf_s,
+        swscanf,
+        swscanf_s,
+        vfprintf,
+        vfprintf_s,
+        vfscanf,
+        vfscanf_s,
+        vfwprintf,
+        vfwprintf_s,
+        vfwscanf,
+        vfwscanf_s,
+        vprintf,
+        vprintf_s,
+        vscanf,
+        vscanf_s,
+        vsnprintf,
+        vsnprintf_s,
+        vsprintf,
+        vsprintf_s,
+        vsscanf,
+        vsscanf_s,
+        vswprintf,
+        vswprintf_s,
+        vswscanf,
+        vswscanf_s,
+        vwprintf,
+        vwprintf_s,
+        vwscanf,
+        vwscanf_s,
+        wprintf,
+        wprintf_s,
+        wscanf,
+        wscanf_s,
+
+        // time
+        gmtime
+        ;
+
+        private static final Map<String, WindowsFallbackSymbols> BY_NAME;
+
+        static {
```

Comment:
Wouldn't Enum.valueOf achieve the same result here?

Reply:
Right, I'm so used to the static map idiom for when the key is something else than the name that I didn't really think about using Enum.valueOf.

I'll switch to using valueOf.